presto-spi/src/test/java/com/facebook/presto/spi/block/BenchmarkFloorMod.java

```diff
@@ -0,0 +1,185 @@
+        }
+    }
+
+    @Benchmark
+    public void floorMod(RandBenchmarkData data, Blackhole blackhole)
+            throws Throwable
+    {
+        long[] hashcodes = data.getHashcodes();
+        for (long i : hashcodes) {
+            blackhole.consume(Math.floorMod(i, hashTableSize));
+        }
+    }
+
+    @SuppressWarnings("FieldMayBeFinal")
+    @State(Scope.Thread)
+    public static class RandBenchmarkData
+    {
+        private Random random;
+        long[] hashcodes = new long[numHashCodes];
+
+        @Setup
+        public void setup()
+                throws Exception
+        {
+            random = new Random(0);
+            for (int i = 0; i < numHashCodes; i++) {
+                hashcodes[i] = random.nextLong();
+            }
+        }
+
+        private long[] getHashcodes()
+        {
+            return hashcodes;
+        }
+    }
+
+    private int getReducedHashPosition1(long hashcode, int hashTableSize)
+    {
+        return (int) (hashcode & 0x7fff_ffff_ffff_ffffL) % hashTableSize;
+    }
+
+    private int getReducedHashPosition2(long hashcode, int hashTableSize)
+    {
+        hashcode = hashcode & 0x7fff_ffff_ffff_ffffL;
+        return (int) (hashcode - hashcode / hashTableSize * hashTableSize);
+    }
+
+    // This reduction function requires the hashTableSize to be power of 2.
+    private int getReducedHashPosition3(long hashcode, int hashTableSize)
+    {
+        return (int) hashcode & (hashTableSize - 1);
+    }
+
+    // Adapted from the fastrange algorithm: https://github.com/lemire/fastrange. This function reduces the 64 bit hashcode
+    // to [0, hashTableSize) uniformly. It first gets a 32 bit hashcode from the lower 32 bit of the 64 bit hashcode,
+    // then reduce this 32 bit hash code to [0, hashTableSize) by multiplying it by hashTableSize then divided by 2^32.
+    // By multiplying by hashTableSize, we take 32 bit hashcode in the range [0, 2^32) and map them to multiples of
+    // hashTableSize in [0, hashTableSize * 2^32). By dividing by 2^32, we map all multiples of hashTableSize in [0, 2^32)
+    // to 0, all multiples of hashTableSize in [2^32, 2 * 2^32) to one, and so forth. In each interval of 2^32 size, the
+    // number of reduced values must be 2^32/hashTableSize, therefore the distribution is uniform if the hashcode
+    // distribution is uniform. This is essentially calculating hashTableSize * (32-bit hashcode / 2^32).
+    private static int getReducedHashPosition4(long hashcode, int hashTableSize)
+    {
+        return (int) (((hashcode & 0xffff_ffffL) * (long) hashTableSize) >> 32);
+    }
+
+    // This function is similar to getReducedHashPosition4, just changing the way of reducing the 64 bit hashcode to 32
+    // bit by calling Long.hashCode() in front. Long.hashCode() returns the XOR of the high 32 bits and low 32 bits. But
+    // we still need to and it with 0xffff_ffffL because when it's negative the high 32 bits would be all 1's.
```

Comment:
do we still need to & with `0xffff_ffffL` for this `reduceHashPosition` method?

Reply:
yes. Long.hashCode() could be negative. And before multiplying with hashTableSize which is a long type, it would be cast to long and become something like 0xffff_ffff_ff08_ac50

Reply:
What about writing the last sentence in this way:
```
... x & 0xffff_ffffL turns x (treated as unsigned integer)  into long. 
```

Reply:
Actually, we can use `Integer.toUnsignedLong` which is self-explained. Thanks @haozhun